### PR TITLE
c8d: Fix image commit with userns mapping (carry)

### PR DESCRIPTION
**Pull request by dmcgowan**
Alternative approach to #46938 as mentioned [here](https://github.com/moby/moby/pull/46938#issuecomment-1865249630).

### CODE REVIEW

**Comment by dmcgowan** [comment]
Marking this one as ready to review, I don't see any problems in the tests and any optimizations can come later if necessary.

**Comment by thaJeztah** [comment]
Looks like there's a compile failure on Windows;

```
..\..\daemon\containerd\image_commit.go:265:28: undefined: remapSuffix
..\..\daemon\containerd\image_commit.go:284:16: i.copyAndUnremapRootFS undefined (type *ImageService has no field or method copyAndUnremapRootFS)
```

File: daemon/containerd/image_snapshot_unix.go [hunk 1]
```diff
@@ -85,7 +110,48 @@ func (i *ImageService) unremapRootFS(ctx context.Context, mounts []mount.Mount)
...
+
+	for _, xattr := range xattrKeys {
+		data, err := sysx.LGetxattr(path, xattr)
+		if err != nil {
+			return err
+		}
+		xattrs[xattr] = data
+	}
+
+	if err := os.Lchown(path, uid, gid); err != nil {
```
Comment (thaJeztah) [inline daemon/containerd/image_snapshot_unix.go:137]: Can be in a follow-up, but considering to add some details about why these attributes are stripped in the first place (I was initially scratching my head on that), and why we have to go against that. Perhaps it's considered "common knowledge" but I wasn't aware (🙈) Djordje posted a link to https://lwn.net/Articles/244747/ on Slack, which made that clear to me.

File: daemon/containerd/image_snapshot_windows.go [hunk 2]
```diff
@@ -7,6 +7,14 @@ import (
 	"github.com/containerd/containerd/snapshots"
 )
 
+var (
+	remapSuffix = "-remap"
+)
```
Comment (thaJeztah) [inline daemon/containerd/image_snapshot_windows.go:12]: nit; looks like this changed to a var in the windows variant; could you make this a `const` here as well? (and while doing so make it a single line, to keep `gofumpt` happy 😅)

```suggestion
const remapSuffix = "-remap"
```

File: daemon/containerd/image_snapshot_unix.go (hunk 1 above)
Comment (dmcgowan) [reply daemon/containerd/image_snapshot_unix.go:137]: That sounds useful and would help me understand it as well.

**Review by rumpl — APPROVED** [review]
LGTM :+1:

**Review by thaJeztah — APPROVED** [review]
LGTM let's bring this one in

**Pull request merged**